src/cluster-client/src/metrics.rs

```diff
@@ -0,0 +1,134 @@
+                name: "mz_dataflow_wallclock_lag_seconds_count",
+                help: "The total count of dataflow wallclock lag measurements.",
+                var_labels: ["instance_id", "replica_id", "collection_id"],
+            )),
+        }
+    }
+
+    /// Return an object that tracks wallclock lag metrics for the given collection on the given
+    /// cluster and replica.
+    pub fn wallclock_lag_metrics(
+        &self,
+        collection_id: String,
+        instance_id: Option<String>,
+        replica_id: Option<String>,
+    ) -> WallclockLagMetrics {
+        let labels = vec![
+            instance_id.unwrap_or_default(),
+            replica_id.unwrap_or_default(),
+            collection_id.to_string(),
+        ];
+
+        let labels_with_quantile = |quantile: &str| {
+            labels
+                .iter()
+                .cloned()
+                .chain([quantile.to_string()])
+                .collect()
+        };
+
+        let wallclock_lag_seconds_min = self
+            .dataflow_wallclock_lag_seconds
+            .get_delete_on_drop_metric(labels_with_quantile("0"));
+        let wallclock_lag_seconds_max = self
+            .dataflow_wallclock_lag_seconds
+            .get_delete_on_drop_metric(labels_with_quantile("1"));
+        let wallclock_lag_seconds_sum = self
+            .dataflow_wallclock_lag_seconds_sum
+            .get_delete_on_drop_metric(labels.clone());
+        let wallclock_lag_seconds_count = self
+            .dataflow_wallclock_lag_seconds_count
+            .get_delete_on_drop_metric(labels);
+        let wallclock_lag_minmax = SlidingMinMax::new(60);
+
+        WallclockLagMetrics {
+            wallclock_lag_seconds_min,
+            wallclock_lag_seconds_max,
+            wallclock_lag_seconds_sum,
+            wallclock_lag_seconds_count,
+            wallclock_lag_minmax,
+        }
+    }
+}
+
+/// Metrics tracking frontier wallclock lag for a collection.
+#[derive(Debug)]
+pub struct WallclockLagMetrics {
+    /// Gauge tracking minimum dataflow wallclock lag.
+    wallclock_lag_seconds_min: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
+    /// Gauge tracking maximum dataflow wallclock lag.
+    wallclock_lag_seconds_max: DeleteOnDropGauge<'static, AtomicF64, Vec<String>>,
+    /// Counter tracking the total sum of dataflow wallclock lag.
+    wallclock_lag_seconds_sum: DeleteOnDropCounter<'static, AtomicF64, Vec<String>>,
+    /// Counter tracking the total count of dataflow wallclock lag measurements.
+    wallclock_lag_seconds_count: DeleteOnDropCounter<'static, AtomicU64, Vec<String>>,
+
+    /// State maintaining minimum and maximum wallclock lag.
+    wallclock_lag_minmax: SlidingMinMax<f32>,
+}
+
+impl WallclockLagMetrics {
+    /// Observe a new wallclock lag measurement.
+    pub fn observe(&mut self, lag: Duration) {
+        let lag_secs = lag.as_secs_f32();
+
+        self.wallclock_lag_minmax.add_sample(lag_secs);
```

Comment:
To my untrained eye this state seems to be storing exactly the same info that `wallclock_lag_seconds_{min,max}` are storing. Why do we need to store them twice?

Reply:
`wallclock_lag_minmax` is the sliding min/max over the last minute (really: over the last 60 samples, where we mean to collect a sample every second). Whereas `wallclock_lag_seconds_{min,max}` are the Prometheus metrics that only store a single value, they cannot be used to calculate a sliding window. We want to Prometheus metrics to always reflect the min/max over the last minute, and we maintain `wallclock_lag_minmax` to achieve that.